### PR TITLE
General cleanup in validate.go

### PR DESCRIPTION
**Pull request by vrindle**
Uses more expressive return types and also reduces cyclomatic complexity.

### CODE REVIEW

**Comment by bn222** [comment]
As discussed, let's have `validateNicModel` return an error instead. Also, let's get rid of the side-effect way to communicate the result of functions. More precisely, we should at least address `interfaceSelected` and maybe even `nodesSelected`. 

It seems to me that we should also refactor the for loop in the function dynamicValidateSriovNetworkNodePolicy. Some parts of it are nested 5 deep.

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:318]: This is divergent logic. The "for" loop is trying to find something matching. It doesn't mean that if it doesn't match that it is an error.

So the new behaviour of the code you have implemented here is not equivalent to the original.

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:229]: Err contains important information. But it is not passed down or shown to the user.

Comment (vrindle) [reply pkg/webhook/validate.go]: This can be resolved by logging and showing to the user without changing the behavior too much. Latest changes reflect that.

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:213]: Is `interfaceSelected` variable used? If not, then remove the variable up top.

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:206]: Shouldn't be pass node by pointer and not value? E.g. `node *corev1.Node` like in `validatePolicyForNodeState()`

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:250]: Shouldn't we also fix up this function as well? It looks like we always return "true, nil" or "false, error string"

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:242]: So what is the purpose of `validateNicModel`? Is it a function that we call with every interface to see if the policy matches with it? Is it really an error if we don't match?

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:308]: Isn't it also important to log that the vendor/deviceID is not supported?

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:282]: I don't think we should be considering all these cases as errors. Rather these cases are matching criteria. For example if an Intel NIC being checked does not equal a Nvidia NIC, then it is considered verbose logging to debug an issue. To see which NICs didn't match with the policy.

File: pkg/webhook/validate.go
Comment (vrindle) [reply pkg/webhook/validate.go:308]: We do log that. This returns true which means that it is supported. When it reaches it reaches the end of the function we log that vendor/device ID is not supported. https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/422/files#diff-8fd5d83d7413d77ee2eab8e5c1a2ea1623d63e2aec69017a95811db9ad9edf52R319

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:240]: So the logic before was if any interface selected was set to true (i.e. validateNicModel returned true) then the  print `fmt.Errorf("no supported NIC is selected by the nicSelector in CR %s", cr.GetName())` won't be activated. However with this new logic, we will be verbosely logging every interface that doesn't match. Do you have a sample log of this that you can share? And it would be apparent.

So we need to somehow store all the debug information going through all the interfaces such that when  `validateNicModel` never returned true then we will give the user all the debug information because nothing matched. This is to improve the verbosity of `no supported NIC is selected by the nicSelector in CR %s` print and not to add verbosity overall.

@bn222  Is this your understanding as well?

File: pkg/webhook/validate_test.go
Comment (wizhaoredhat) [inline pkg/webhook/validate_test.go:216]: Test coverage is no longer complete because you have new error messages.

**Comment by SchSeba** [comment]
Hi @vrindle can you please fix the commit message?

the tittle is longer then expected for a commit :)

**Comment by SchSeba** [comment]
@vrindle please also add a test to cover the new case where the nic is supported but not initialize and please check the error message

File: pkg/webhook/validate.go
Comment (bn222) [reply pkg/webhook/validate.go:213]: This comment should be resolved

File: pkg/webhook/validate.go
Comment (bn222) [reply pkg/webhook/validate.go:246]: +1, we need separate commits in this PR. Cleanup PRs that do not change behavior + one commit that actually adds the new error message that we wanted to add in the first place.

File: pkg/webhook/validate.go
Comment (bn222) [reply pkg/webhook/validate.go:240]: yes, One approach (and please do give your feedback on this) is to call an additional function that only logs things. It would log which nics it tried and why it didn't work out.

Comment (wizhaoredhat) [reply pkg/webhook/validate.go:240]: I think I am on the same page as you.

File: pkg/webhook/validate_test.go
Comment (vrindle) [reply pkg/webhook/validate_test.go:216]: Since we only adding errors for debugging/logging information, the test coverage that tests errors should be complete.

File: pkg/webhook/validate.go
Comment (bn222) [reply pkg/webhook/validate.go:244]: @vrindle Can you separate out the commits? `git rebase -i` and `git add -p`

File: pkg/webhook/validate_test.go
Comment (wizhaoredhat) [reply pkg/webhook/validate_test.go:216]: We should test that the debugging/logging information is what we expect based on certain parameters.

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [reply pkg/webhook/validate.go]: @vrindle Removing this print does not solve the issue. There needs to be additional logic to consolidate the information and present it only in the case when no interfaces match.

Comment (vrindle) [reply pkg/webhook/validate.go]: Right I am implementing that in a second commit. This one commit is only to ensure that the general cleanup is done.

Comment (bn222) [reply pkg/webhook/validate.go]: Can we rename the PR so that it's clear that we are doing a cleanup first and do all the refactoring in this first PR. The aim to reduce the size of the PR is that it can then go in. Next, we can have an additional PR after that that only adds the error message we wanted.

Comment (vrindle) [reply pkg/webhook/validate.go]: That sounds good. I renamed the PR. This one only has the cleanup commit. Then there is this PR: https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/434

Please ensure this PR gets merged before merging the one above since this is a stacked PR.

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:213]: Cleanup should only mean refactoring. You are changing the behaviour of the code here. We want this error to be printed when no interface is found.

Comment (vrindle) [reply pkg/webhook/validate.go]: Okay, so I reverted this to what it was before. You can find the error when no interface is found to be in the other PR.

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:173]: This shouldn't be removed. It is still needed.

File: pkg/webhook/validate.go
Comment (bn222) [inline pkg/webhook/validate.go:30]: Wondering why we still have this here. Isn't the goal of the cleanup to not pass the value of this bool through a global variable? Instead, return it from the function?

Comment (vrindle) [reply pkg/webhook/validate.go:30]: This is general cleanup. Earlier you requested this to be split into 2 PRs so that is what I did. This one needs to merged before the other PR gets merged.

Comment (bn222) [reply pkg/webhook/validate.go:30]: ok, sounds good then. Thanks!

File: pkg/webhook/validate.go
Comment (wizhaoredhat) [inline pkg/webhook/validate.go:194]: We still return bool, err here. Should we clean this up as well?

File: pkg/webhook/validate_test.go
Comment (wizhaoredhat) [inline pkg/webhook/validate_test.go:625]: Don't we still need this?

Comment (wizhaoredhat) [reply pkg/webhook/validate_test.go:625]: You removed `interfaceSelected` from the tests when it is still used.

**Comment by vrindle** [comment]
@bn222 
@wizhaoredhat 
@zshi-redhat

File: pkg/webhook/validate_test.go
Comment (wizhaoredhat) [inline pkg/webhook/validate_test.go:677]: Should we add more tests for `validateNicModel`, since now it returns an error instead of a boolean?

Comment (vrindle) [reply pkg/webhook/validate_test.go:677]: what sort of tests are you looking to add? Imo, these functions test validateNicModel quite heavily.

Comment (wizhaoredhat) [reply pkg/webhook/validate_test.go:677]: `validatePolicyForNodeState` gets tested. But in `err := validateNicModel(&policy.Spec.NicSelector, &iface, node)`, the `err` gets thrown away. I think it would be worth it to test `validateNicModel` indepedently.

Comment (vrindle) [reply pkg/webhook/validate_test.go:677]: I agree that we should test that out but maybe not in this PR. In my other PR I  test this when the variable interfaceSelected is removed. Please see this line here: https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/434/files#diff-ae978883f1787164fb6fa85b31c5fde6421f8a06d9c9f5f013fb72e6f21863a0R560

I think we can leave the tests as is for now, and then we can add this test in the new PR.

Comment (wizhaoredhat) [reply pkg/webhook/validate_test.go:677]: Ok

**Comment by wizhaoredhat** [comment]
LGTM

**Comment by bn222** [comment]
@adrianchiris 

This is a pretty small change that doesn't actually change functionality. CI is green and we already have an LGTM from RH. Can you PTAL? Should be pretty quick.

Regards,
Balazs

**Comment by adrianchiris** [comment]
> This is a pretty small change that doesn't actually change functionality. CI is green and we already have an LGTM from RH. 

did not see an LGTM from maintainer, @bn222 @SchSeba you should approve :)
will take a look at this later this week.

**Comment by bn222** [comment]
lgtm

File: pkg/webhook/validate.go
Comment (adrianchiris) [inline pkg/webhook/validate.go:231]: nit:

```
if err := validateNicModel; err  == nil {
  ...
}
```

that said, its really a matter of taste here so im fine with keeping as is.

**Review by adrianchiris — APPROVED** [review]

**Comment by adrianchiris** [comment]
Thx @vrindle for this improvement !

**Pull request merged**
